File: linked_list/polinomial.cpp
#include <stdio.h>
#include <stdlib.h>

typedef struct Hang{
    int coef;
    int seung;
    Hang* next;
}Hang;

void createHang(Hang** node, int coef, int seung);
void printHang(Hang* node);
void sumHang(Hang* f, Hang* g, Hang** result);
void chaHang(Hang* f, Hang* g, Hang** result);
void attachHang(int coef, Hang* f, Hang** attach);
void removeHang(Hang** node);

int main(){
    Hang* fx = NULL;
    Hang* gx = NULL;
    Hang* sum = NULL;
    Hang* cha = NULL;

    createHang(&fx, 5, 12);
    createHang(&fx, -3, 9);
    createHang(&fx, 2, 7);
    createHang(&fx, -10, 5);
    createHang(&fx, 5, 4);
    createHang(&fx, -8, 3);
    createHang(&fx, 6, 2);
    createHang(&fx, -4, 1);
    createHang(&fx, 10, 0);
    printf("f(x) = ");
    printHang(fx);

    createHang(&gx, 13, 11);
    createHang(&gx, 7, 10);
    createHang(&gx, 5, 9);
    createHang(&gx, 3, 8);
    createHang(&gx, -4, 7);
    createHang(&gx, 1, 6);
    createHang(&gx, -12, 5);
    createHang(&gx, 9, 4);
    createHang(&gx, -5, 3);
    createHang(&gx, 1, 2);
    createHang(&gx, 7, 1);
    createHang(&gx, -23, 0);
    printf("g(x) = ");
    printHang(gx);

    sumHang(fx, gx, &sum);
    printf("f(x) + g(x) = ");
    printHang(sum);

    chaHang(fx, gx, &cha);
    printf("f(x) - g(x) = ");
    printHang(cha);

    removeHang(&fx);
    removeHang(&gx);
    removeHang(&sum);
    removeHang(&cha);
    return 0;
}

void createHang(Hang** node, int coef, int seung){
    if(*node!=NULL){
        return createHang(&((*node)->next), coef, seung);
    }
    *node = (Hang *)malloc(sizeof(Hang));
    (*node)->coef = coef;
    (*node)->seung = seung;
    (*node)->next = NULL;
}

void printHang(Hang* node){
    if(node==NULL){
        printf("\n");
        return;
    }
    if(node->coef==1) printf("+");
    else if(node->coef==-1) printf("-");
    else printf("%+d", node->coef);
    if(node->seung!=0) printf("x^%d", node->seung);
    return printHang(node->next);
}

void sumHang(Hang* f, Hang* g, Hang** result){
    if(f==NULL && g==NULL) return;
    if(f==NULL) return attachHang(1, g, result);
    if(g==NULL) return attachHang(1, f, result);
    if(f->seung > g->seung){
        createHang(result, f->coef, f->seung);
        sumHang(f->next, g, result);
    } else if (f->seung < g->seung){
        createHang(result, g->coef, g->seung);
        sumHang(f, g->next, result);
    } else {
        createHang(result, f->coef + g->coef, f->seung);
        sumHang(f->next, g->next, result);
    }
}

void chaHang(Hang* f, Hang* g, Hang** result){
    if(f==NULL && g==NULL) return;
    if(f==NULL) return attachHang(-1, g, result);
    if(g==NULL) return attachHang(1, f, result);
    if(f->seung > g->seung){
        createHang(result, f->coef, f->seung);
        chaHang(f->next, g, result);
    } else if (f->seung < g->seung){
        createHang(result, (-1) * g->coef, g->seung);
        chaHang(f, g->next, result);
    } else {
        createHang(result, f->coef - g->coef, f->seung);
        chaHang(f->next, g->next, result);
    }
}

void attachHang(int coef, Hang* f, Hang** attach){
    if(f==NULL) return;
    createHang(attach, coef * f->coef, f->seung);
    attachHang(coef, f->next, attach);
}

void removeHang(Hang** node){
    if(*node){
        removeHang(&((*node)->next));
        free(*node);
        *node = NULL;
    }
}

File: linked_list/polinomial.py
class Hang:
    def __init__(self, coef, exp):
        self.coef = coef
        self.exp = exp
        self.next = None

    def addHang(self, coef, exp):
        if self.next == None:
            self.next = Hang(coef, exp)
        else:
            self.next.addHang(coef, exp)

    def printHang(self):
        if self.coef > 0:
            print("+", end="")
        print("{}x^{}".format(self.coef, self.exp), end="")
        if self.next != None:
            self.next.printHang()
        else:
            print()

def sumPol(f, g):
    if(f.exp > g.exp):
        result = Hang(f.coef, f.exp)
        f = f.next
    elif(f.exp < g.exp):
        result = Hang(g.coef, g.exp)
        g = g.next
    else:
        result = Hang(f.coef + g.coef, f.exp)
        f = f.next
        g = g.next
    while f!=None and g!=None:
        if (f.exp > g.exp):
            result.addHang(f.coef, f.exp)
            f = f.next
        elif (f.exp < g.exp):
            result.addHang(g.coef, g.exp)
            g = g.next
        else:
            result.addHang(f.coef + g.coef, f.exp)
            f = f.next
            g = g.next
    while f!=None:
        result.addHang(f.coef, f.exp)
        f = f.next
    while g != None:
        result.addHang(g.coef, g.exp)
        g = g.next
    return result

def chaPol(f, g):
    if (f.exp > g.exp):
        result = Hang(f.coef, f.exp)
        f = f.next
    elif (f.exp < g.exp):
        result = Hang((-1) * g.coef, g.exp)
        g = g.next
    else:
        result = Hang(f.coef - g.coef, f.exp)
        f = f.next
        g = g.next
    while f != None and g != None:
        if (f.exp > g.exp):
            result.addHang(f.coef, f.exp)
            f = f.next
        elif (f.exp < g.exp):
            result.addHang((-1)*g.coef, g.exp)
            g = g.next
        else:
            result.addHang(f.coef - g.coef, f.exp)
            f = f.next
            g = g.next
    while f != None:
        result.addHang(f.coef, f.exp)
        f = f.next
    while g != None:
        result.addHang((-1)*g.coef, g.exp)
        g = g.next
    return result

fx = Hang(5, 12)
fx.addHang(-3, 9)
fx.addHang(2, 7)
fx.addHang(-10, 5)
fx.addHang(5, 4)
fx.addHang(-8, 3)
fx.addHang(6, 2)
fx.addHang(-4, 1)
fx.addHang(10, 0)

print("f(x) = ", end="")
fx.printHang()

gx = Hang(13, 11)
gx.addHang(7, 10)
gx.addHang(5, 9)
gx.addHang(3, 8)
gx.addHang(-4, 7)
gx.addHang(1, 6)
gx.addHang(-12, 5)
gx.addHang(9, 4)
gx.addHang(-5, 3)
gx.addHang(1, 2)
gx.addHang(7, 1)
gx.addHang(-23, 0)

print("g(x) = ", end="")
gx.printHang()

print("f(x) + g(x) = ", end="")
sumPol(fx, gx).printHang()

print("f(x) - g(x) = ", end="")
chaPol(fx, gx).printHang()

input()